modules/tensor_mechanics/src/materials/ComputePeriodicStrain.C

```diff
@@ -0,0 +1,81 @@
+#include "RankTwoTensor.h"
+
+registerMooseObject("TensorMechanicsApp", ComputePeriodicStrain);
+
+template <>
+InputParameters
+validParams<ComputePeriodicStrain>()
+{
+  InputParameters params = validParams<Material>();
+  params.addClassDescription(
+      "Material for storing the periodic strain values from the scalar variable");
+  params.addParam<std::string>("base_name",
+                               "Optional parameter that allows the user to define "
+                               "multiple mechanics material systems on the same "
+                               "block, i.e. for multiple phases");
+  params.addCoupledVar("scalar_periodic_strain", "Scalar variable for periodic strain");
+  return params;
+}
+
+ComputePeriodicStrain::ComputePeriodicStrain(const InputParameters & parameters)
+  : Material(parameters),
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _scalar_periodic_strain(coupledScalarValue("scalar_periodic_strain")),
+    _scalar_periodic_strain_order(coupledScalarOrder("scalar_periodic_strain")),
+    _periodic_strain(declareProperty<RankTwoTensor>(_base_name + "periodic_strain"))
+{
+}
+
+void
+ComputePeriodicStrain::initQpStatefulProperties()
+{
+  _periodic_strain[_qp].zero();
+}
+
+void
+ComputePeriodicStrain::computeProperties()
+{
+  RankTwoTensor strain;
+  if (_scalar_periodic_strain_order == FIRST)
+    strain = RankTwoTensor(_scalar_periodic_strain[0], 0, 0, 0, 0, 0);
+  else if (_scalar_periodic_strain_order == THIRD)
+    strain = RankTwoTensor(_scalar_periodic_strain[0],
+                           _scalar_periodic_strain[1],
+                           0,
+                           0,
+                           0,
+                           _scalar_periodic_strain[2]);
+  else if (_scalar_periodic_strain_order == SIXTH)
+    strain = RankTwoTensor(_scalar_periodic_strain[0],
+                           _scalar_periodic_strain[1],
+                           _scalar_periodic_strain[2],
+                           _scalar_periodic_strain[3],
+                           _scalar_periodic_strain[4],
+                           _scalar_periodic_strain[5]);
+  else if (_scalar_periodic_strain_order == NINTH)
+    strain = RankTwoTensor(_scalar_periodic_strain[0],
+                           _scalar_periodic_strain[1],
+                           _scalar_periodic_strain[2],
+                           _scalar_periodic_strain[3],
+                           _scalar_periodic_strain[4],
+                           _scalar_periodic_strain[5],
+                           _scalar_periodic_strain[6],
+                           _scalar_periodic_strain[7],
+                           _scalar_periodic_strain[8]);
+  else
+    mooseError("ComputePerdiodicStrain is only compatible with FIRST, THIRD, SIXTH,"
+               "and NINTH order scalar variables.");
+
+  for (_qp = 0; _qp < _qrule->n_points(); ++_qp)
+    _periodic_strain[_qp] = strain;
+}
```

Comment:
since this material is constant over the domain we might be able to speed this up even further...